### ISSUE TITLE
Can I verify the `can` and `pay` performance?

### ISSUE THREAD

**Issue by ArchyZheng** [RishiHazra/saycanpay#5]
Hi,

I would like to ask whether you have experienced any unusual phenomena with your model's learning curve, specifically a sudden spike towards the end of the training epochs when I train `can` and `pay` model.

<img width="324" alt="image" src="https://github.com/RishiHazra/saycanpay/assets/120232918/ea756130-cfd7-4ff7-8e12-30183e6f74e2">

something like this.

Could you tell me whether you guys meet this problem?

**Comment by RishiHazra** [comment]
Hi @ArchyZheng, yes.
I used the train/validation curve to pick a stable checkpoint (or restrict the training to fewer epochs). I believe it converges to a degenerate solution with more training -- perhaps a better loss function (or even regularization) could have addressed that but I didn't look much into it.

**Comment by ArchyZheng** [comment]
Hi @RishiHazra,

Thank you for your kindly reply. I just fix this problem, by disable the callback function.
https://github.com/RishiHazra/saycanpay/blob/1a395d56e33894e1761c826e441172378458fb10/planLM/train/babyai_train.py#L52

If you have time, maybe you can try to verify my problem. And this is a great paper! But I still have some problem, as you mention in your paper your `Can` module is a probability about `pre(a_t)`. But I don't know how to understand this representation with your `Can` module implementation.

At last, thanks you for your inspiriting idea behind the paper.🫡

**Comment by RishiHazra** [comment]
> Hi @RishiHazra,
> 
> Thank you for your kindly reply. I just fix this problem, by disable the callback function.
> 
> https://github.com/RishiHazra/saycanpay/blob/1a395d56e33894e1761c826e441172378458fb10/planLM/train/babyai_train.py#L52
> 
> If you have time, maybe you can try to verify my problem. And this is a great paper! But I still have some problem, as you mention in your paper your `Can` module is a probability about `pre(a_t)`. But I don't know how to understand this representation with your `Can` module implementation.
> 
> At last, thanks you for your inspiriting idea behind the paper.🫡

Appreciate the fix. I'll definitely try it.
As for the can model: the underlying idea is to learn whether an action can be performed in a given state. This is similar to modeling preconditions in planning (preconditions are used to indicate whether an action can be executed in a given state, however while preconditions are [explicitly specified for planning](https://planning.wiki/ref/pddl/domain), in saycanpay, we're trying to learn them) -- also similar to affordances in robotics. I hope that provides some clarity. But feel free to write back if not. Cheers!

**Comment by RishiHazra** [comment]
Just to add to the previous reply, the InfoNCE loss teaches the Can model feasible (positive) and infeasible (negative) actions for a given state. So implicitly, it's learning what actions are possible in the given state.

**Comment by ArchyZheng** [comment]
Thank you! I can understand the idea more clearly now!
